src/util/merkle_set.py

```diff
@@ -68,129 +71,172 @@ def compress_root(mystr):
         self.root = self.root.remove(toremove, 0)
 
-    def is_included_already_hashed(self, tocheck):
+    def is_included_already_hashed(self, tocheck: bytes) -> Tuple[bool, bytes]:
         proof: List = []
         r = self.root.is_included(tocheck, 0, proof)
         return r, b"".join(proof)
 
-    def _audit(self, hashes):
+    def _audit(self, hashes: List[bytes]):
         newhashes: List = []
         self.root._audit(newhashes, [])
         assert newhashes == sorted(newhashes)
 
 
-class EmptyNode:
+class EmptyNode(Node):
     def __init__(self):
         self.hash = BLANK
 
-    def get_hash(self):
+    def get_hash(self) -> bytes:
         return EMPTY + BLANK
 
-    def is_empty(self):
+    def is_empty(self) -> bool:
         return True
 
-    def is_terminal(self):
+    def is_terminal(self) -> bool:
         return False
 
     def is_double(self):
         raise SetError()
 
-    def add(self, toadd, depth):
+    def add(self, toadd: bytes, depth: int) -> Node:
         return TerminalNode(toadd)
 
-    def remove(self, toremove, depth):
+    def remove(self, toremove: bytes, depth: int) -> Node:
         return self
 
-    def is_included(self, tocheck, depth, p):
+    def is_included(self, tocheck: bytes, depth: int, p: List[bytes]) -> bool:
         p.append(EMPTY)
         return False
 
-    def other_included(self, tocheck, depth, p, collapse):
+    def other_included(self, tocheck: bytes, depth: int, p: List[bytes], collapse: bool):
         p.append(EMPTY)
 
-    def _audit(self, hashes, bits):
+    def _audit(self, hashes: List[bytes], bits: List[int]):
         pass
 
 
 _empty = EmptyNode()
 
 
-class TerminalNode:
-    def __init__(self, hash, bits=None):
+class TerminalNode(Node):
+    def __init__(self, hash: bytes, bits: List[int] = None):
         assert len(hash) == 32
         self.hash = hash
         if bits is not None:
             self._audit([], bits)
 
-    def get_hash(self):
+    def get_hash(self) -> bytes:
         return TERMINAL + self.hash
 
-    def is_empty(self):
+    def is_empty(self) -> bool:
         return False
 
-    def is_terminal(self):
+    def is_terminal(self) -> bool:
         return True
 
-    def is_double(self):
+    def is_double(self) -> bool:
         raise SetError()
 
-    def add(self, toadd, depth):
+    def add(self, toadd: bytes, depth: int) -> Node:
         if toadd == self.hash:
             return self
         if toadd > self.hash:
             return self._make_middle([self, TerminalNode(toadd)], depth)
         else:
             return self._make_middle([TerminalNode(toadd), self], depth)
 
-    def _make_middle(self, children, depth):
+    def _make_middle(self, children: Any, depth: int) -> Node:
         cbits = [get_bit(child.hash, depth) for child in children]
         if cbits[0] != cbits[1]:
             return MiddleNode(children)
-        nextvals = [None, None]
+        nextvals: List[Node] = [_empty, _empty]
```

Comment:
I'm not 100% sure about this. I did this to make mypy happy. Probably I can also remove the next line.

Reply:
I think that's fine, they are both getting overriden.